### ISSUE TITLE
Allow updating node pools using ARM template

### ISSUE THREAD

**Issue by aelij** [Azure/AKS#2194]
**What happened**:
Currently the only option to add a node pool in an ARM template is by creating a separate child resource (`.../providers/Microsoft.ContainerService/managedClusters/aks1/agentPools/p2`). This presents a problem when trying to apply an update to the primary (system) node pool which requires recreating it, for example, to allow it to join an existing subnet or enable "encryption at host".

If we were to add the new agent pool using a child resource in the template, the template will no longer be idempotent (i.e. it won't be able to deploy a new clean environment) and also the template would not clean up the old pool. It forces us to use scripts to complement ARM.

ARM deployments were made to be idempotent and this essentially breaks it.

**What you expected to happen**:
Allow updating node pools using the `agentPoolProfiles` array of the `managedClusters` type.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.18.14
- Size of cluster (how many worker nodes are in the cluster?) 5
- General description of workloads in the cluster: HTTP microservices

**Comment by kelly-brown** [comment]
I agree this really needs to be done.  Put the idea out on the feedback site.  Any way we could get an ETA or find out if this is even planned?

https://feedback.azure.com/forums/914020-azure-kubernetes-service-aks/suggestions/43214700-enable-nodepool-image-updates-via-arm

**Comment by danimal521** [comment]
@palma21 can you look into this one?

**Comment by aelij** [comment]
Not stale :)

**Comment by aelij** [comment]
Still not stale

**Comment by aelij** [comment]
Not stale

**Comment by tspearconquest** [comment]
This also affects the azurerm terraform provider.
Seems to be an antipattern to have the nodepools as a separate resource rather than a block inside the cluster's resource definition block.

**Comment by aelij** [comment]
@tspearconquest I don't think the problem is that it's a child resource, it's the duality - it's both a child resource and an array property. And you have to keep them in sync.
If I could choose, I'd go for something that allows creating a cluster with NO node pools and add the pools exclusively using child resources. This allows for more granular updates.

**Comment by tspearconquest** [comment]
Apologies, I didn't realize there was a nuance there that I was missing in my phrasing. Thank you.

On the Terraform side, there is a "default_node_pool" block which is part of the parent resource. Trying to do anything to that nodepool in terraform code, or making any changes to it manually, and then running terraform, can end up destroying and recreating your cluster. This concept of a "default" node pool doesn't seem to be defined anywhere in Azure, while conversely "system" and "user" node pools are well known to me. I agree with you regarding creating a cluster with no nodepools, which is why I brought this up.

It definitely seems strange to have a default nodepool definition in the cluster resource definition (for terraform) in the first place, given that there is no "default" in Azure. In fact, I was just able to delete a default nodepool using azure-cli the other day and re-link a secondary nodepool I had setup as the "default" nodepool in my terraform state file without incurring any downtime.

Of course, that's why I'm here. It was an all day affair take care of it, all in order to change 2 settings that can only be modified when a nodepool (or cluster) is created.

**Comment by fschmied** [comment]
Here's the error message given when you try to add a node pool from the parent resource's ARM deployment. (Adding it so this is easier to find from the Internet.)
```
The template deployment 'aksCluster' is not valid according to the validation procedure. The tracking id is '...'. See inner errors for details. (Code:
     | InvalidTemplateDeployment)  - Provisioning of resource(s) for container service ... in resource group ... failed. Message: Category: ClientError; Code: BadRequest; SubCode: ; Message: A new agent pool was introduced. Adding agent pools to an existing cluster is
     | not allowed through managed cluster operations. For agent pool specific change, please use per agent pool operations: https://aka.ms/agent-pool-rest-api;
```

**Comment by aelij** [comment]
Not stale

**Comment by acortelyou** [comment]
Not stale

**Comment by stack111** [comment]
> @palma21 can you look into this one?

@palma21 any chance we could have some response for this issue?

**Comment by denniszielke** [comment]
@kaarthis would this be something for the upgrade node pool scenario? It would certainly make upgrading easier.

**Comment by fschmied** [comment]
Not stale.

**Comment by fschmied** [comment]
Not stale.

**Comment by jaumecen** [comment]
Facing this issue right now, 2 years and still nothing. It's a hedache trying to update the cluster system node and the user ones at the same time automatically with an ARM :/

**Comment by fschmied** [comment]
Not stale

**Comment by tspearconquest** [comment]
not stale

**Comment by tspearconquest** [comment]
not stale

**Comment by matthchr** [comment]
We're aware of this problem and discussing solutions internally. I don't have an ETA for when a fix will be forthcoming but we're very aware that this breaks certain use-cases. cc @phealy, @bmoore-msft

**Comment by jaumecen** [comment]
not stale

**Comment by tspearconquest** [comment]
not stale

**Comment by fschmied** [comment]
not stale

**Comment by matthchr** [comment]
not stale

**Comment by fschmied** [comment]
Not stale. This would be incredibly useful.

**Comment by fschmied** [comment]
Not stale

**Comment by fschmied** [comment]
Not stale.

**Comment by fschmied** [comment]
Still relevant

**Comment by fschmied** [comment]
@matthchr Any news, perhaps? This would be so useful... 

And can the issue perhaps be exempt from the staleness checks somehow, as long as you're still discussing it?

**Comment by matthchr** [comment]
I think I've silenced the staleness check.
I dont have much to share on this right now, @phealy is driving some internal efforts related to this but I dont think we have a concrete ETA yet.

**Comment by fschmied** [comment]
Just FYI - for everyone using Terraform's AzureRM provider, this has been worked around in v4.16.0. See https://github.com/hashicorp/terraform-provider-azurerm/issues/22265 .

(It's still an issue for people using ARM or Bicep, of course.)
